compiler/rustc_mir/src/transform/simplify.rs

```diff
@@ -477,7 +487,21 @@ fn remove_unused_definitions<'a, 'tcx>(used_locals: &'a mut UsedLocals, body: &m
                     StatementKind::StorageLive(local) | StatementKind::StorageDead(local) => {
                         used_locals.is_used(*local)
                     }
-                    StatementKind::Assign(box (place, _)) => used_locals.is_used(place.local),
+                    StatementKind::Assign(box (place, _)) => {
+                        let used = used_locals.is_used(place.local);
+                        let mut is_zst = false;
+
+                        // ZST locals can be removed
+                        if used && !is_static {
```

Comment:
We refactored this loop a few months ago to push most of this kind of logic into `UsedLocals`. i think that's probably the place we need to make this change. 

It might just be a matter of adjusting this logic:

https://github.com/rust-lang/rust/blob/e2267046859c9ceb932abc983561d53a117089f6/compiler/rustc_mir/src/transform/simplify.rs#L392-L398

Reply:
Took a stab at moving the logic into `UsedLocals` with the lastest commits